comm/src/main/scala/io/casperlabs/comm/gossiping/DownloadManagerImpl.scala

```
@@ -153,18 +156,36 @@ class DownloadManagerImpl[F[_]: Sync: Concurrent](
     backend.hasBlock(hash)
 
   // TODO: Just say which block hash to download, try all possible sources.
-  def download(summary: BlockSummary, source: Node, relay: Boolean): F[Unit] =
-    for {
-      block <- downloadAndRestore(source, summary.blockHash)
-      // TODO: Validate
-      // TODO: Store
-      // TODO: Gossip
-      // TODO: Signal failure.
-      _ <- signal.put(Signal.Downloaded(summary.blockHash))
-    } yield ()
+  def download(summary: BlockSummary, source: Node, relay: Boolean): F[Unit] = {
+    val id = Base16.encode(summary.blockHash.toByteArray)
+    // Try to download until we succeed.
+    def loop(): F[Unit] = {
+      val tryDownload =
+        for {
+          // TODO: Pick a source.
+          block <- downloadAndRestore(source, summary.blockHash)
+          // TODO: Validate
+          // TODO: Store
+          // TODO: Gossip
+        } yield ()
+
+      tryDownload.recoverWith {
+        case ex =>
+          // TODO: Exponential backoff, pick another node, try to store again, etc.
+          Log[F].error(s"Failed to download block $id", ex)
```

So are you basically ignoring any errors that occur while downloading?

---

I changed it later so that the last error is returned to all callers, if there is no more possible recourse (which will be a separate ticket).